### PR TITLE
Enable bitsandbytes packaging for ROCm

### PR DESCRIPTION
**Pull request by pnunna93**
This PR enables packaging wheels for bitsandbytes on ROCm. It updates rocm compilation and wheels build jobs to compile on ROCm 6.1 for latest Instinct and Radeon GPUs.

There are also updates to documentation based on user feedback and steps to install bitsandbytes ROCm package once available.

cc: @Titus-von-Koeller @sunway513 @amathews-amd

### CODE REVIEW

**Comment by Titus-von-Koeller** [comment]
This looks very good! From my perspective I don't see anything amiss. @matthewdouglas could you take a second look, just to be sure and then we merge?

cc @akx if you have a moment and are interested to take a look :) we always highly appreciate your opinion!

File: bitsandbytes/cextension.py [hunk 1]
```diff
@@ -119,8 +119,10 @@ def get_native_library() -> BNBNativeLibrary:
     if torch.version.hip:
         hip_major, hip_minor = map(int, torch.version.hip.split(".")[0:2])
         HIP_ENVIRONMENT, BNB_HIP_VERSION = True, hip_major * 100 + hip_minor
+        BNB_HIP_VERSION_SHORT = str(hip_major) + str(hip_minor)
```
Comment (akx) [inline bitsandbytes/cextension.py:122]:
```suggestion
        BNB_HIP_VERSION_SHORT = f"{hip_major}{hip_minor}"
```

**Review by akx — CHANGES_REQUESTED** [review]
Beyond the small suggestion about string formatting, should the naming be unified? What's HIP c.f. ROCm? :)

(Oh, and either the commit history here should be cleaned, or this should be squash-merged :) )

**Comment by pnunna93** [comment]
> Beyond the small suggestion about string formatting, should the naming be unified? What's HIP c.f. ROCm? :)
> 
> (Oh, and either the commit history here should be cleaned, or this should be squash-merged :) )

Thanks for the suggestion! HIP is a wrapper around ROCm, it supports both ROCm and CUDA platforms. The major/minor versions are same for HIP and ROCm.

File: docs/source/installation.mdx [hunk 2]
```diff
@@ -146,15 +146,25 @@ Please follow these steps to install bitsandbytes with device-specific backend s
...
-# Create a docker container with latest pytorch. It comes with ROCm and pytorch preinstalled
-docker pull rocm/pytorch:latest
-docker run -it --device=/dev/kfd --device=/dev/dri --group-add video rocm/pytorch:latest
+# Create a docker container with latest ROCm image, which includes ROCm libraries
+docker pull rocm/dev-ubuntu-22.04:6.1.2-complete
+docker run -it --device=/dev/kfd --device=/dev/dri --group-add video rocm/dev-ubuntu-22.04:6.1.2-complete
+apt-get update && apt-get install -y git && cd home
 
+# Install pytorch compatible with above ROCm version
+pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/rocm6.1/
```
Comment (matthewdouglas) [inline docs/source/installation.mdx:158]: I think it would make sense to use the 2.4.0 release, unless for some reason that does not work?

Comment (pnunna93) [reply docs/source/installation.mdx]: It works, I just removed --pre option so that 2.4.0 is installed.

**Review by matthewdouglas — APPROVED** [review]

File: docs/source/installation.mdx [hunk 3]
```diff
@@ -146,15 +146,25 @@ Please follow these steps to install bitsandbytes with device-specific backend s
...
-# Create a docker container with latest pytorch. It comes with ROCm and pytorch preinstalled
-docker pull rocm/pytorch:latest
-docker run -it --device=/dev/kfd --device=/dev/dri --group-add video rocm/pytorch:latest
+# Create a docker container with latest ROCm image, which includes ROCm libraries
+docker pull rocm/dev-ubuntu-22.04:6.1.2-complete
+docker run -it --device=/dev/kfd --device=/dev/dri --group-add video rocm/dev-ubuntu-22.04:6.1.2-complete
+apt-get update && apt-get install -y git && cd home
 
+# Install pytorch compatible with above ROCm version
+pip install torch --index-url https://download.pytorch.org/whl/nightly/rocm6.1/
```
Comment (Titus-von-Koeller) [inline docs/source/installation.mdx:158]:
```suggestion
pip install torch --index-url https://download.pytorch.org/whl/rocm6.1/
```

**Review by Titus-von-Koeller — CHANGES_REQUESTED** [review]

**Comment by Titus-von-Koeller** [comment]
Thanks @pnunna93 great work and thanks a lot @akx and @matthewdouglas for helping with the review 🙌🏻

**Pull request merged**